devcareer/code-jammers-backend
# #175355965 Build CRUD API for state

## What does this PR do ##
-  enables user to get one state with specific id 
-  enables user to get all states
- enables admin to edit an existing state in the database
- enables admin to delete a state from the database
- enables admin to add a state to the database

### Description of tasks to be completed? ###
- create get single state route
- create get multiple states routes
- create get single state controller
- create get multiple states controller
- create a delete state controller
- create a delete state route
- create an edit state controller
- create an edit state route
- write tests for the controllers and routes

### How should this be manually tested/checked? ###
- Clone the PR
- Checkout to the branch `ft-user-should-get-state-175355965`
- Install the dependencies using `npm install`
- Create a `.env file` and define the following as follows:
   - `DEV_DATABASE_URL = <development database url>`
   - `TEST_DATABASE_URL = <test database url>`
   - `DATABASE_URL = <production database url>`
- run ```npm run migrate``` to create the database tables
- run ```npm run seed``` to seed the database
- Start the server by running ```npm run start-dev```
- Sign in via http://localhost:3000/api/v1/users/signin using a POST method with
          {
          "email": "ufuoma@gmail.com",
          "password": "12345"
           }
and copy the token (without the quotes) from the response body.

#### Add State ####
- In a new tab, under the Authorization tab, change the TYPE to Bearer Token and paste the token in the space available. Make a POST request to `http://localhost:3000/api/v1/admin/state/:countryId` to add a state. place the following in the request body and send to add the state.
```
{
    "name": "oyo",
    "capital": "ibadan",
    "gallery":
            "https://en.wikipedia.org/wiki/Lusaka#/media/File:Downtown_Lusaka.JPG"

}
```
#### Get all States ####
- In a new tab, Make a `GET` request to `http://localhost:3000/api/v1/states` to get all states
copy any state id from the response data labeled as `id`
#### Get a State ####
- In a new tab, Make a `GET` request to `http://localhost:3000/api/v1/state/<paste copied id>` to get the selected state
#### UPDATE a State ####
- In a new tab, Make a `PATCH` request to `http://localhost:3000/api/v1/admin/state/<paste copied id>` and place the fields to be updated in the request body.
    - request body sample:
```
{
    "countryId": "2e11e4a9-441b-4426-9521-39adc64ccfad",
    "name": "asaba",
    "capital": "delta",
    "gallery": "https://en.wikipedia.org/wiki/Lusaka#/media/File:Downtown_Lusaka.JPG"
}
```
#### DELETE a State ####
- In a new tab, Make a `DELETE` request to `http://localhost:3000/api/v1/admin/state/<paste copied id>` to delete the selected state.

#### State associations ####
-  make GET a request to url endpoint `http://localhost:3000/api/v1/countries`
it should return all the countries with their associated tourist centers and states in the database.



### Link to architecture
[architecture](https://docs.google.com/document/d/1zCepZxQ8aH6UYuvadjRtpgpSjLWfGLcNZHPnZPpAszM/edit?usp=sharing)

### What are the relevant pivotal tracker stories?
[#175355965](https://www.pivotaltracker.com/story/show/175355965)
